### DISCUSSION TITLE
Build from source (and run) in MacOS (CMake build)

### DISCUSSION THREAD

**Discussion by carajillu** [cp2k/cp2k#3556] [Q&A] [answered]
I am trying to build CP2K in MacOS (blame my work) using the CMake build. I have managed to build it but then when I try to run a small calculation, I get the following error
```shell
jclarknicholas@M-5ce91e87c9cd cp2k_test % ~/local/cp2k/bin/cp2k.psmp -i h2o.in 
[M-5ce91e87c9cd:28136] shmem: mmap: an error occurred while determining whether or not /var/folders/n3/6vxjv07d3lbfbcgw0ql6dcw80000gr/T//ompi.M-5ce91e87c9cd.504/jf.0/4059168768/sm_segment.M-5ce91e87c9cd.504.f1f20000.0 could be created.
At line 360 of file /Users/jclarknicholas/github/cp2k/build/src/base/machine.F (unit = 121245)
Fortran runtime error: Cannot open file '/proc/self/statm': No such file or directory

Error termination. Backtrace:

Could not print backtrace: executable file is not an executable

Could not print backtrace: executable file is not an executable

Could not print backtrace: executable file is not an executable

Could not print backtrace: executable file is not an executable

Could not print backtrace: executable file is not an executable
#0  0x10345ed2f
#1  0x10345f8d7
#2  0x1034601a7
#3  0x10354f0a3
#4  0x10354f5bf
#5  0x107a6c267
#6  0x1066d124f
#7  0x1027b881f
#8  0x1027ba99f
``` 

According to chatGPT, this error is related to MPI and happens because /proc doesn't exist in MacOS, so I tried installing CP2K with -DCP2K_USE_MPI=off, and when I run it this is what I get (trying to use MPI?). It does build the ssmp executable and not the psmp, so not sure why this happens
```shell
jclarknicholas@M-5ce91e87c9cd cp2k_test % ~/local/cp2k/bin/cp2k.ssmp -i h2o.in 
*** The MPI_Comm_f2c() function was called before MPI_INIT was invoked.
*** This is disallowed by the MPI standard.
*** Your MPI job will now abort.
[M-5ce91e87c9cd:53034] Local abort before MPI_INIT completed completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
jclarknicholas@M-5ce91e87c9cd cp2k_test %
``` 

Below are the commands I use to configure and build:
```shell
cmake .. -DCMAKE_INSTALL_PREFIX=/Users/jclarknicholas/local/cp2k -DCMAKE_PREFIX_PATH="/Users/jclarknicholas/local/dbcsr;/opt/homebrew/Cellar/lapack/3.12.0/;/Users/jclarknicholas/local/scalapack-2.2.1/;/opt/homebrew/Cellar/openblas/0.3.27/" -DCMAKE_C_COMPILER=gcc -DCMAKE_Fortran_COMPILER=gfortran
make -j8 install
export DYLD_LIBRARY_PATH=/Users/jclarknicholas/local/cp2k/lib/
``` 

Anybody seen a similar problem before?

Thanks!

**Comment by mkrack** [comment]
In case, you do not insist on a cmake build, you may have a look at this [How-to Compile CP2K on macOS](https://www.cp2k.org/howto:compile_on_macos).

**Comment by carajillu** [comment]
Thanks @mkrack for the suggestion. I've installed it using the toolchain just to rule out a problem with my laptop, and it works, however, I believe for development, the CMake build makes things a lot easier for me. 

So, if the toolchain build works, that means that somehow the functions in src/base/machine.F that cause trouble are not being called. I believe they have to do with memory management, so something similar must be being used by the toolchain build. The question is: what is it, and how do I tell CMake to use it instead of machine.F? 

I've been trying to understand the differences between the MacOS and Linux files in arch/, but having a lot of trouble with it, so any ideas, anybody?

Thanks a lot!

**Comment by carajillu** [comment, marked as answer]
So! 
After a lot of fiddling, I finally made it work! The trick is to add this when you call CMake:
```shell
-DCMAKE_Fortran_FLAGS="-D__NO_STATM_ACCESS"
```
I'll mark this as answer but I'll leave it open in case somebody tries and finds other problems.

**Reply by carajillu** [reply]
Forgot to mention, this error still appears, but the calculation runs (I don't need scientifically valid results for now so it's good enough for me)
```shell
shmem: mmap: an error occurred while determining whether or not /var/folders/n3/6vxjv07d3lbfbcgw0ql6dcw80000gr/T//ompi.M-5ce91e87c9cd.504/jf.0/4059168768/sm_segment.M-5ce91e87c9cd.504.f1f20000.0 could be created.
```
